network/network-address/src/lib.rs

```diff
@@ -0,0 +1,619 @@
+/// `&str`.
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+#[serde(try_from = "String", into = "String")]
+pub struct DnsName(String);
+
+/// Possible errors when parsing a human-readable [`NetworkAddress`].
+#[derive(Error, Debug)]
+pub enum ParseError {
+    #[error("unknown protocol type: '{0}'")]
+    UnknownProtocolType(String),
+
+    #[error("unexpected end of protocol string")]
+    UnexpectedEnd,
+
+    #[error("error parsing ip4/ip6 address: {0}")]
+    ParseAddrError(#[from] net::AddrParseError),
+
+    #[error("error parsing integer: {0}")]
+    ParseIntError(#[from] num::ParseIntError),
+
+    #[error("error parsing x25519 public key: {0}")]
+    ParseX25519PubkeyError(#[from] CryptoMaterialError),
+
+    #[error("network address cannot be empty")]
+    EmptyProtocolString,
+
+    #[error("protocol string must start with '/'")]
+    InvalidProtocolString,
+
+    #[error("dns name cannot be empty")]
+    EmptyDnsNameString,
+
+    #[error("dns name cannot contain '/' characters")]
+    InvalidDnsNameCharacter,
+
+    #[error("dns name is too long: len: {0} bytes, max len: 255 bytes")]
+    DnsNameTooLong(usize),
+}
+
+#[derive(Error, Debug)]
+#[error("network address cannot be empty")]
+pub struct EmptyError;
+
+///////////////////////
+// RawNetworkAddress //
+///////////////////////
+
+impl RawNetworkAddress {
+    pub fn new(bytes: Vec<u8>) -> Self {
+        Self(bytes)
+    }
+
+    pub fn empty() -> Self {
+        Self(Vec::new())
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+}
+
+impl Into<Vec<u8>> for RawNetworkAddress {
+    fn into(self) -> Vec<u8> {
+        self.0
+    }
+}
+
+impl AsRef<[u8]> for RawNetworkAddress {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl TryFrom<&NetworkAddress> for RawNetworkAddress {
+    type Error = lcs::Error;
+
+    fn try_from(value: &NetworkAddress) -> Result<Self, lcs::Error> {
+        let bytes = lcs::to_bytes(value)?;
+        Ok(RawNetworkAddress::new(bytes))
+    }
+}
+
+////////////////////
+// NetworkAddress //
+////////////////////
+
+impl NetworkAddress {
+    fn new(protocols: Vec<Protocol>) -> Self {
+        Self(protocols)
+    }
+}
+
+impl FromStr for NetworkAddress {
+    type Err = ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.is_empty() {
+            return Err(ParseError::EmptyProtocolString);
+        }
+
+        let mut protocols = Vec::new();
+        let mut parts_iter = s.split('/');
+
+        // the first character must be '/'
+        if parts_iter.next() != Some("") {
+            return Err(ParseError::InvalidProtocolString);
+        }
+
+        // parse all `Protocol`s
+        while let Some(protocol_type) = parts_iter.next() {
+            protocols.push(Protocol::parse(protocol_type, &mut parts_iter)?);
```

Comment:
so you can have multiple times the same protocol, or have unordered protocols. Why have all protocols be part of the same enum? Why not have an enum per /X/ ?